pkg/gameserverallocations/controller.go

```diff
@@ -493,44 +509,154 @@ func (c *Controller) serialisation(r *http.Request, w http.ResponseWriter, obj k
 	return errors.Wrapf(err, "error encoding %T", obj)
 }
 
-// allocate allocated a GameServer from a given Fleet
+// allocate allocated a GameServer from a given GameServerAllocation
+// this sets up allocation through a batch process.
 func (c *Controller) allocate(gsa *v1alpha1.GameServerAllocation) (*stablev1alpha1.GameServer, error) {
-	var allocation *stablev1alpha1.GameServer
-	var comparator findComparator
+	req := request{gsa: gsa, response: make(chan response)}
 
-	switch gsa.Spec.Scheduling {
-	case apis.Packed:
-		comparator = packedComparator
-	case apis.Distributed:
-		comparator = distributedComparator
+	c.pendingRequests <- req
+
+	select {
+	case res := <-req.response:
+		return res.gs, res.err
+	case <-c.stop:
+		return nil, errors.New("shutting down")
 	}
+}
 
-	allocation, err := c.findReadyGameServerForAllocation(gsa, comparator)
-	if err != nil {
-		return allocation, err
+// runLocalAllocations is a blocking function that runs in a loop
+// looking at c.requestBatches for batches of requests that are coming through.
+func (c *Controller) runLocalAllocations(updateWorkerCount int) {
+	updateQueue := make(chan response)
+	// setup workers for allocation updates
+	c.allocationUpdateWorkers(updateQueue, updateWorkerCount)
+
+	var list []*stablev1alpha1.GameServer
+	requestCount := 0
+
+	for {
+		select {
+		case req := <-c.pendingRequests:
+			// refresh the list after every 100 allocations made in a single batch
+			requestCount++
+			if requestCount >= maxBatchBeforeRefresh {
+				list = nil
+				requestCount = 0
+			}
+
+			if list == nil {
+				list = c.listSortedReadyGameServers()
+			}
+
+			gs, index, err := findGameServerForAllocation(req.gsa, list)
+			if err != nil {
+				req.response <- response{request: req, gs: nil, err: err}
+				continue
+			}
+			// remove the game server that has been allocated
+			list = append(list[:index], list[index+1:]...)
+
+			key, _ := cache.MetaNamespaceKeyFunc(gs)
+			if ok := c.readyGameServers.Delete(key); !ok {
+				// this seems unlikely, but lets handle it just in case
+				req.response <- response{request: req, gs: nil, err: ErrConflictInGameServerSelection}
+				continue
+			}
+
+			updateQueue <- response{request: req, gs: gs.DeepCopy(), err: nil}
+
+		case <-c.stop:
+			return
+		default:
+			list = nil
+			requestCount = 0
+			// slow down cpu churn, and allow items to batch
+			time.Sleep(batchWaitTime)
```

Comment:
/cc @jkowalski @roberthbailey @ilkercelikyilmaz 

Curious to see your take on the Sleep approach above. I wanted to avoid nested select statements as much as possible, to keep the code as simple/understandable as I could (since it's lots of async processing) - and this seemed like the best way forward.

Would love your feedback!

Reply:
When queue depth is less than 100 (maxBatchQueueSize) and we don't have any more requests, how long will the allocations in the queue will wait?

Reply:
The longest they will wait <= `batchWaitTime` (500ms) because we might be in the sleep code)  - but that will only be for the first request. After that, each request in the batch (c.pendingRequests) is processed immediately.